LIBCLOUD 473 Add EBS Volume Attributes

cderamus opened the pull request:
https://issues.apache.org/jira/browse/LIBCLOUD-473

cderamus committed:
Merge pull request #11 from apache/trunk

Synchronizing branch

Commit:
Issue LIBCLOUD-473: Add EBS Volume Attributes

Commit:
Merge branch 'LIBCLOUD-473_Add_EBS_Volume_Attributes' of https://github.com/DivvyCloud/libcloud into LIBCLOUD-473_Add_EBS_Volume_Attributes

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -876,19 +876,89 @@ def _to_volume(self, element, name):
+
+        # Set our name if the Name key/value if available
+        # If we don't get anything back then use the volume id
+        name = tags.get('Name', volId)
```
This now shadows `name` argument which is passed to this function.

Maybe a better approach would be to use `name` argument  which is passed to the function (if available), otherwise fall back to retrieving a name from the tags (`name = name if name else tags.get('Name', volId)).

Kami replied:
While you are at it, I would also do the following changes:
1. Modify `_to_volume` method signature and default `name` to `None`
2. Modify `self._to_volume(el,``)` call to omit name argument (empty string) all together
3. Add a docstring which explains that if not provided, name is obtained from the tags (if available)

cderamus replied:
Not a problem, will do.

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -876,19 +876,89 @@ def _to_volume(self, element, name):
-                          namespace=NAMESPACE)
+
+        # Get our tag items
+        tag_items = findall(element=element,
```
There seem to be multiple places in the file where the tags are read into the dict and exactly the same code is repeated (`ex_describe_tags`, `_to_node`, `_to_network`, etc.)

It would be great if this functionality could be refactored in a separate function and re-used.

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -876,19 +876,89 @@ def _to_volume(self, element, name):
+                             namespace=NAMESPACE)
+            if value is not None:
+                # Convert our create/attach time to ISO 8601
+                if attribute == 'create_time' or attribute == 'attach_time':
```
Maybe instead of `type`, you could call it a `format_func` / `cast_func`, specify `parse_date` as `format_func` for those two attributes and get rid of this if.

cderamus replied:
Sounds good.

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -876,19 +876,89 @@ def _to_volume(self, element, name):
+
+        # Define and build our extra dictionary
+        extra = {}
+        for attribute, values in extra_attributes_map.items():
```
This functionality now also appears to be repeated multiple times in this file.

It would be great if you could refactor this into a separate function (e.g. `_extract_extra_attributes` or something like that) which takes extraction definition (basically, `extra_attributes_map` argument) and returns a `dict` with extracted attributes.

cderamus replied:
I was thinking about doing that actually, but wanted to submit it as a separate issue for tracking purposes. Once this PR is closed I'll go ahead and create a JIRA issue for that and go through all calls where the extra_attributes_map exists if you're okay with that approach.

Kami replied:
Yeah, I'm fine with that.

Commit:
Added new function _get_resource_tags to build and return the tags dictionary
Changed _to_volume to default name to None
Updated the extra_attributes_map to cast_func and removed an unnecessary if condition

cderamus commented:
Assuming that you're okay with these changes I will create a new branch later today to convert all extra_attributes_map dictionaries from type -> cast_func to remain consistent and also will set up the new utility function that we previously discussed. Thanks!

Kami commented:
@cderamus Yeah, sounds good.

I've squashed the commits, updated the docstring and merged patch into trunk. Thanks!
